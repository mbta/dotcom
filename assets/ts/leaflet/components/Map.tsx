import React, { ReactElement, useLayoutEffect } from "react";
import { uniqBy } from "lodash";
import deepEqual from "fast-deep-equal/react";
import {
  Icon,
  LatLngBounds,
  Marker as LeafletMarker,
  MarkerOptions
} from "leaflet";
import Leaflet from "react-leaflet";
import { MapData, MapMarker, IconOpts } from "./__mapdata";

export interface ZoomOpts {
  maxZoom: number;
  minZoom: number;
  scrollWheelZoom: boolean;
  style: object;
}

export const defaultZoomOpts: ZoomOpts = {
  maxZoom: 18,
  minZoom: 9,
  scrollWheelZoom: false,
  style: { touchAction: "none" } // fix for chrome-android to prevent zooming page
};

interface Props {
  bounds?: LatLngBounds;
  boundsByMarkers?: boolean;
  mapData: MapData;
}

const mapCenter = (
  markers: MapMarker[],
  { latitude, longitude }: { latitude: number; longitude: number }
): [number, number] | undefined => {
  if (markers.length === 1) return [markers[0].latitude, markers[0].longitude];
  return [latitude, longitude];
};

const rotateMarker = (
  markerEl: LeafletMarker<MarkerOptions>,
  { rotation_angle: angle }: MapMarker
): void => {
  // TS does not recognize setRotationAngle, which
  // is added by the leaflet-rotatedmarker plugin
  // @ts-ignore
  markerEl.setRotationAngle(angle);
};

const Component = ({
  bounds,
  boundsByMarkers,
  mapData: {
    default_center: defaultCenter,
    markers,
    polylines,
    tile_server_url: tileServerUrl,
    zoom
  }
}: Props): ReactElement<HTMLElement> | null => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const mapRef = React.createRef<any>();
  useLayoutEffect(() => {
    if (mapRef.current && mapRef.current.leafletElement) {
      // tells Leaflet to redraw the map
      mapRef.current.leafletElement.invalidateSize();
    }
  }, [mapRef]);

  if (typeof window !== "undefined" && tileServerUrl !== "") {
    /* eslint-disable */
    const leaflet: typeof Leaflet = require("react-leaflet");
    const buildIcon: (
      icon: string | null,
      opts?: IconOpts
    ) => Icon | undefined = require("../icon").default;
    require("leaflet-rotatedmarker");
    const getBounds = require("../bounds").default;
    /* eslint-enable */
    const { Map, Marker, Polyline, Popup, TileLayer } = leaflet;
    /* istanbul ignore next */
    const boundsOrByMarkers = bounds || (boundsByMarkers && getBounds(markers));
    const position = mapCenter(markers, defaultCenter);
    const nonNullZoom = zoom === null ? undefined : zoom;

    const onTilesLoaded = (): void => {
      mapRef.current.container.parentElement.classList += " map--loaded";
    };

    return (
      <Map
        ref={mapRef}
        bounds={boundsOrByMarkers}
        center={position}
        zoom={nonNullZoom}
        {...defaultZoomOpts}
      >
        <TileLayer
          attribution='&amp;copy <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          url={`${tileServerUrl}/osm_tiles/{z}/{x}/{y}.png`}
          onload={onTilesLoaded}
        />
        {polylines.map(
          /* istanbul ignore next */
          polyline => (
            <Polyline
              className={polyline.className}
              key={
                polyline.id || `polyline-${Math.floor(Math.random() * 1000)}`
              }
              positions={polyline.positions}
              color={polyline.color}
              weight={polyline.weight}
              dashArray={polyline["dotted?"] ? "4px 8px" : "none"}
              lineCap={polyline["dotted?"] ? "butt" : "round"}
              lineJoin={polyline["dotted?"] ? "miter" : "round"}
            />
          )
        )}
        {uniqBy(markers, "id").map(marker => (
          <Marker
            icon={buildIcon(marker.icon, marker.icon_opts)}
            key={
              marker.id ||
              /* istanbul ignore next */ `marker-${Math.floor(
                Math.random() * 1000
              )}`
            }
            alt={marker.alt || "Marker"}
            position={[marker.latitude, marker.longitude]}
            ref={ref => ref && rotateMarker(ref.leafletElement, marker)}
            zIndexOffset={marker.z_index}
            keyboard={false}
            onclick={marker.onClick}
          >
            {marker.tooltip && <Popup offset={[0, 40]}>{marker.tooltip}</Popup>}
          </Marker>
        ))}
      </Map>
    );
  }
  return null;
};

export default React.memo(Component, deepEqual);
