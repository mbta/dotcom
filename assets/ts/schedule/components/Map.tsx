import React, { ReactElement, Reducer, useRef } from "react";
import { SocketEvent } from "../../app/channels";
import LeafletMap from "../../leaflet/components/Map";
import getBounds from "../../leaflet/bounds";
import {
  MapData,
  MapMarker as Marker
} from "../../leaflet/components/__mapdata";
import CrowdingPill from "./line-diagram/CrowdingPill";
import useChannel from "../../hooks/useChannel";

interface Props {
  channel: string;
  data: MapData;
  currentShapes: string[];
  currentStops: string[];
}

export const iconOpts = (
  icon: string | null
): {
  icon_size?: [number, number];
  icon_anchor?: [number, number];
} => {
  switch (icon) {
    case null:
      return {};

    case "vehicle-bordered-expanded":
      return {
        icon_size: [18, 18], // eslint-disable-line camelcase
        icon_anchor: [6, 6] // eslint-disable-line camelcase
      };

    case "stop-circle-bordered-expanded":
      return {
        icon_size: [12, 12], // eslint-disable-line camelcase
        icon_anchor: [6, 6] // eslint-disable-line camelcase
      };

    default:
      throw new Error(`unexpected icon type: ${icon}`);
  }
};

const zIndex = (icon: string | null): number | undefined =>
  icon === "vehicle-bordered-expanded" ? 1000 : undefined;

const updateMarker = (marker: Marker): Marker => ({
  ...marker,
  tooltip: (
    <div>
      {marker.vehicle_crowding && (
        <>
          <CrowdingPill crowding={marker.vehicle_crowding} />
          <br />
        </>
      )}
      {marker.tooltip_text}
    </div>
  ),
  icon_opts: iconOpts(marker.icon), // eslint-disable-line camelcase
  z_index: zIndex(marker.icon) // eslint-disable-line camelcase
});

const isVehicleMarker = (marker: Marker): boolean =>
  marker.icon ? marker.icon.includes("vehicle") : false;

interface IdHash {
  [id: string]: true;
}

const shouldRemoveMarker = (id: string | null, idHash: IdHash): boolean =>
  id !== null && idHash[id] === true;

interface EventData {
  marker: Marker;
}

type DataFromSocketType = SocketEvent<EventData[]>;
interface DataReducerState {
  markers: Marker[];
}
type DataReducerAction = DataFromSocketType;
type DataReducerType = Reducer<DataReducerState, DataReducerAction>;

export const reducer: DataReducerType = (state, action) => {
  switch (action.event) {
    case "reset":
      return {
        ...state,
        markers: state.markers
          .filter(marker => !isVehicleMarker(marker))
          .concat(
            action.data.map(({ marker }: EventData) => updateMarker(marker))
          )
      };

    case "add":
      return {
        ...state,
        markers: state.markers.concat(
          action.data.map(({ marker }) => updateMarker(marker))
        )
      };

    case "update":
      if (action.data.length === 0) {
        return state;
      }
      // Filter out the existing marker if necessary, always add new marker
      return {
        ...state,
        markers: [
          updateMarker(action.data[0].marker),
          ...state.markers.filter(
            marker => marker.id !== action.data[0].marker.id
          )
        ]
      };
    case "remove":
      return {
        ...state,
        markers: state.markers.filter(
          marker =>
            !shouldRemoveMarker(
              marker.id,
              action.data.reduce((acc: IdHash, id: string) => {
                acc[id] = true;
                return acc;
              }, {})
            )
        )
      };
    default:
      /* istanbul ignore next */
      // eslint-disable-next-line no-console
      console.warn("unexpected event", action);
      return state;
  }
};

const Map = ({
  data,
  channel,
  currentShapes,
  currentStops
}: Props): ReactElement<HTMLElement> | null => {
  const state = useChannel<DataFromSocketType, DataReducerType>(
    channel,
    reducer,
    { markers: [] }
  );

  const stopMarkers = data.stop_markers
    ? data.stop_markers
        .filter(mark => currentStops.includes(mark.id as string))
        .map(marker => updateMarker(marker))
    : [];

  const mapData = {
    ...data,
    polylines: data.polylines.filter(p =>
      currentShapes.some(shape => shape === p.id)
    ),
    markers: state.markers.concat(stopMarkers)
  };
  const bounds = useRef(getBounds(stopMarkers));
  return (
    <div className="m-schedule__map">
      <LeafletMap bounds={bounds.current} mapData={mapData} />
    </div>
  );
};

export default Map;
