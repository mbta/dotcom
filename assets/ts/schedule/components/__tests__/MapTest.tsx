import { mount } from "enzyme";
import React from "react";
import { TileLayer } from "react-leaflet";
import {
  MapData,
  MapMarker as Marker
} from "../../../leaflet/components/__mapdata";
import Map, { iconOpts, reducer } from "../Map";

/* eslint-disable camelcase */
const data: MapData = {
  zoom: 16,
  width: 600,
  tile_server_url: "https://mbta-map-tiles-dev.s3.amazonaws.com",
  polylines: [],
  stop_markers: [
    {
      icon: "stop-circle-bordered-expanded",
      id: "stop-place-alfcl",
      latitude: 42.395428,
      longitude: -71.142483,
      rotation_angle: 0,
      tooltip: null,
      tooltip_text: "Alewife",
      shape_id: "1"
    }
  ],
  markers: [
    {
      icon: "vehicle-bordered-expanded",
      id: "vehicle-R-545CDFC5",
      latitude: 42.39786911010742,
      longitude: -71.13092041015625,
      rotation_angle: 90,
      tooltip_text: "Alewife train is on the way to Alewife",
      tooltip: null
    },
    {
      icon: "stop-circle-bordered-expanded",
      id: "stop-place-alfcl",
      latitude: 42.395428,
      longitude: -71.142483,
      rotation_angle: 0,
      tooltip: null,
      tooltip_text: "Alewife",
      shape_id: "1"
    }
  ],
  height: 600,
  default_center: {
    longitude: -71.05891,
    latitude: 42.360718
  }
};
/* eslint-enable camelcase */

describe("Schedule Map", () => {
  it("renders", () => {
    const wrapper = mount(
      <Map
        data={data}
        channel="vehicles:Red:0"
        currentShapes={["1", "2"]}
        currentStops={["stop-place-alfcl", "22", "33"]}
      />
    );
    expect(() => wrapper.render()).not.toThrow();
  });

  it("renders and matches snapshot", () => {
    const wrapper = mount(
      <Map
        data={data}
        channel="vehicles:Red:0"
        currentShapes={["1", "2"]}
        currentStops={["stop-place-alfcl", "22", "33"]}
      />
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("adds class name after map load", () => {
    const wrapper = mount(
      <Map
        data={data}
        channel="vehicles:Red:0"
        currentShapes={["1", "2"]}
        currentStops={["stop-place-alfcl", "22", "33"]}
      />
    );

    expect(wrapper.html()).not.toContain("map--loaded");

    const onMapLoad: Function = wrapper.find(TileLayer).prop("onload");
    expect(onMapLoad).toBeInstanceOf(Function);

    // simulate the map load event being fired
    onMapLoad();

    expect(wrapper.html()).toContain("map--loaded");
  });
});

describe("reducer", () => {
  const newMarker: Marker = {
    icon: "vehicle-bordered-expanded",
    id: "vehicle-R-545CDFC6",
    latitude: 42.39786911010742,
    longitude: -71.13092041015625,
    // eslint-disable-next-line camelcase
    rotation_angle: 90,
    // eslint-disable-next-line camelcase
    tooltip_text: "Alewife train is on the way to Alewife",
    tooltip: null,
    // eslint-disable-next-line camelcase
    shape_id: "1"
  };

  it("resets markers", () => {
    const result = reducer(
      { markers: data.markers },
      {
        event: "reset",
        data: [{ marker: newMarker }]
      }
    );

    expect(result.markers.map(m => m.id)).toEqual([
      data.markers[1].id,
      newMarker.id
    ]);
  });

  it("adds vehicles", () => {
    const result = reducer(
      { markers: data.markers },
      {
        event: "add",
        data: [{ marker: newMarker }]
      }
    );
    expect(result.markers.map(m => m.id)).toEqual(
      data.markers.map(m => m.id).concat(newMarker.id)
    );
  });

  it("updates markers", () => {
    const result = reducer(
      { markers: data.markers },
      {
        event: "update",
        data: [{ marker: { ...data.markers[0], latitude: 43.0 } }]
      }
    );

    expect(result.markers.map(m => m.id)).toEqual(data.markers.map(m => m.id));
    expect(data.markers[0].latitude).toEqual(42.39786911010742);
    expect(result.markers[0].latitude).toEqual(43.0);
  });

  it("handles empty data actions", () => {
    const result = reducer(
      { markers: data.markers },
      { event: "update", data: [] }
    );

    expect(result.markers).toEqual(data.markers);
  });

  it("removes markers", () => {
    const result = reducer(
      { markers: data.markers },
      {
        event: "remove",
        data: [data.markers[0].id!]
      }
    );

    expect(result.markers.map(m => m.id)).toEqual([data.markers[1].id]);
  });
});

describe("iconOpts", () => {
  it("handles stop markers", () => {
    expect(iconOpts(data.markers[1].icon)).toEqual({
      icon_size: [12, 12], // eslint-disable-line camelcase
      icon_anchor: [6, 6] // eslint-disable-line camelcase
    });
  });

  it("throws an error if it received an unknown icon type", () => {
    expect(() => iconOpts("unknown")).toThrowError();
  });

  it("does not throw error when icon is null", () => {
    expect(iconOpts(null)).toEqual({});
  });
});
