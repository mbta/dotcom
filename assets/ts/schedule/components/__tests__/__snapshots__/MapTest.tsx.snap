// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule Map renders and matches snapshot 1`] = `
"<Map data={{...}} channel=\\"vehicles:Red:0\\" currentShapes={{...}} currentStops={{...}}>
  <div className=\\"m-schedule__map\\">
    <Component bounds={{...}} mapData={{...}}>
      <Map bounds={{...}} center={{...}} zoom={16} maxZoom={18} minZoom={9} scrollWheelZoom={false} style={{...}}>
        <div className={[undefined]} id={[undefined]} style={{...}}>
          <ForwardRef(Leaflet(TileLayer)) attribution=\\"&copy <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors\\" url=\\"https://mbta-map-tiles-dev.s3.amazonaws.com/osm_tiles/{z}/{x}/{y}.png\\" onload={[Function: onTilesLoaded]}>
            <TileLayer attribution=\\"&copy <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors\\" url=\\"https://mbta-map-tiles-dev.s3.amazonaws.com/osm_tiles/{z}/{x}/{y}.png\\" onload={[Function: onTilesLoaded]} leaflet={{...}} />
          </ForwardRef(Leaflet(TileLayer))>
          <ForwardRef(Leaflet(Marker)) icon={{...}} alt=\\"Marker\\" position={{...}} zIndexOffset={[undefined]} keyboard={false} onclick={[undefined]}>
            <Marker icon={{...}} alt=\\"Marker\\" position={{...}} zIndexOffset={[undefined]} keyboard={false} onclick={[undefined]} leaflet={{...}}>
              <ForwardRef(Leaflet(Popup)) offset={{...}}>
                <Popup offset={{...}} leaflet={{...}} pane=\\"popupPane\\" />
              </ForwardRef(Leaflet(Popup))>
            </Marker>
          </ForwardRef(Leaflet(Marker))>
        </div>
      </Map>
    </Component>
  </div>
</Map>"
`;
